compiler/rustc_lint/src/types.rs

```diff
@@ -1358,3 +1360,216 @@ impl<'tcx> LateLintPass<'tcx> for VariantSizeDifferences {
+    /// `AtomicPtr`, `AtomicUsize`, or any of the other integer atomics).
+    ///
+    /// - Passing `Ordering::Acquire` or `Ordering::AcqRel` to
+    ///   `AtomicType::store`.
+    ///
+    /// - Passing `Ordering::Release` or `Ordering::AcqRel` to
+    ///   `AtomicType::load`.
+    ///
+    /// - Passing `Ordering::Relaxed` to `core::sync::atomic::fence` or
+    ///   `core::sync::atomic::compiler_fence`.
+    ///
+    /// - Passing `Ordering::Release` or `Ordering::AcqRel` as the failure
+    ///   ordering for any of `AtomicType::compare_exchange`,
+    ///   `AtomicType::compare_exchange_weak`, or `AtomicType::fetch_update`.
+    ///
+    /// - Passing in a pair of orderings to `AtomicType::compare_exchange`,
+    ///   `AtomicType::compare_exchange_weak`, or `AtomicType::fetch_update`
+    ///   where the failure ordering is stronger than the success ordering.
+    INVALID_ATOMIC_ORDERING,
+    Deny,
+    "usage of invalid atomic ordering in atomic operations and memory fences"
+}
+
+declare_lint_pass!(InvalidAtomicOrdering => [INVALID_ATOMIC_ORDERING]);
+
+impl InvalidAtomicOrdering {
+    fn type_is_atomic(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        const ATOMIC_TYPES: &[Symbol] = &[
+            sym::atomic_bool_type,
+            sym::atomic_ptr_type,
+            sym::atomic_usize_type,
+            sym::atomic_u8_type,
+            sym::atomic_u16_type,
+            sym::atomic_u32_type,
+            sym::atomic_u64_type,
+            sym::atomic_u128_type,
+            sym::atomic_isize_type,
+            sym::atomic_i8_type,
+            sym::atomic_i16_type,
+            sym::atomic_i32_type,
+            sym::atomic_i64_type,
+            sym::atomic_i128_type,
+        ];
+        if let ty::Adt(&ty::AdtDef { did, .. }, _) = cx.typeck_results().expr_ty(expr).kind() {
+            ATOMIC_TYPES.iter().any(|sym| cx.tcx.is_diagnostic_item(*sym, did))
+        } else {
+            false
+        }
+    }
+
+    fn matches_ordering_def_path(cx: &LateContext<'_>, did: DefId, orderings: &[Symbol]) -> bool {
+        orderings.iter().any(|ordering| {
+            cx.match_def_path(did, &[sym::core, sym::sync, sym::atomic, sym::Ordering, *ordering])
+        })
+    }
+
+    fn opt_ordering_defid(cx: &LateContext<'_>, ord_arg: &Expr<'_>) -> Option<DefId> {
+        if let ExprKind::Path(ref ord_qpath) = ord_arg.kind {
+            cx.qpath_res(ord_qpath, ord_arg.hir_id).opt_def_id()
+        } else {
+            None
+        }
+    }
+
+    fn check_atomic_load_store(cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(ref method_path, _, args, _) = &expr.kind;
+            let method = method_path.ident.name;
+            if Self::type_is_atomic(cx, &args[0]);
+            if method == sym::load || method == sym::store;
+            let ordering_arg = if method == sym::load { &args[1] } else { &args[2] };
```

Comment:
diagnostic items instead of `Symbol` equality

Reply:
This one seems tricky because there's not just one `load` method, there's one for AtomicBool::load, one for AtomicU8::load, one for AtomicU16::load, etc.

I had thought that diagnostic items wouldn't work here (it didn't seem to, but maybe I'm mistaken)

Reply:
no, they do have to be unique afaik :thinking: 

I guess using names is fine in that case. Thanks for looking into this.

Reply:
I'm unsure how to solve this but it also seems like less of an issue. I don't think it's worth having atomic_store_u8, atomic_store_u16, etc methods all have unique diagnostic items.

It feels like you kind of want a `rustc_diagnostic_tag` which would be allowed to have multiple things that it applies to.

As is maybe this is fine though. What's the downside of not using rustc_diagnostic_item here?

Reply:
yeah, someting like `rustc_doagnostic_tag` is what I would want here :+1: 

The main issue is when moving stuff in std which breaks lints like this. I think this may also cause issues if a user impls a trait like the following, or at least it's less obvious for people reading the code what's the expected behavior here
```rust
trait Foo {
    fn load(self);
}
```

In this PR you can just keep using `sym::load` for now

Reply:
can you add the following as an ui test?
```rust
// check-pass
use std::sync::atomic::{AtomicUsize, Ordering};

trait Foo {
    fn store(self, ordering: Ordering);
}

impl Foo for AtomicUsize {
    fn store(self, _ordering: Ordering) {
        AtomicUsize::store(&self, 4, Ordering::SeqCst);
    }
}

fn main() {
    let x = AtomicUsize::new(3);
    x.store(Ordering::Acquire);
}
```